### PR TITLE
[SPARK-24923][SQL] Implement v2 CreateTableAsSelect

### PR DESCRIPTION
**Pull request by rdblue**
## What changes were proposed in this pull request?

This adds a v2 implementation for CTAS queries

* Update the SQL parser to parse CREATE queries using multi-part identifiers
* Update `CheckAnalysis` to validate partitioning references with the CTAS query schema
* Add `CreateTableAsSelect` v2 logical plan and `CreateTableAsSelectExec` v2 physical plan
* Update create conversion from `CreateTableAsSelectStatement` to support the new v2 logical plan
* Update `DataSourceV2Strategy` to convert v2 CTAS logical plan to the new physical plan
* Add `findNestedField` to `StructType` to support reference validation

## How was this patch tested?

We have been running these changes in production for several months. Also:

* Add a test suite `CreateTablePartitioningValidationSuite` for new analysis checks
* Add a test suite for v2 SQL, `DataSourceV2SQLSuite`
* Update catalyst `DDLParserSuite` to use multi-part identifiers (`Seq[String]`)
* Add test cases to `PlanResolutionSuite` for v2 CTAS: known catalog and v2 source implementation

### CODE REVIEW

**Comment by rdblue** [comment]
@mccheah, @cloud-fan, here's a v2 implementation for CTAS. Please review if you have time. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala [hunk 1]
```diff
@@ -46,14 +57,22 @@ case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with Ca
...
       val tableDesc = buildCatalogTable(table, new StructType, partitionCols, bucketSpec,
         properties, provider, options, location, comment, ifNotExists)
       val mode = if (ifNotExists) SaveMode.Ignore else SaveMode.ErrorIfExists
 
       CreateTable(tableDesc, mode, Some(query))
+
+    case create: CreateTableAsSelectStatement =>
+      // the provider was not a v1 source, convert to a v2 plan
+      val CatalogObjectIdentifier(maybeCatalog, identifier) = create.tableName
+      val catalog = maybeCatalog.orElse(defaultCatalog)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: The ability to change the default catalog can be done later. I'd like to add it when we migrate hive catalog to `TableCatalog` API, so that the default can be hive catalog.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala [hunk 2]
```diff
@@ -46,14 +57,22 @@ case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with Ca
...
+        AsTableIdentifier(table), query, partitionCols, bucketSpec, properties,
+        V1WriteProvider(provider), options, location, comment, ifNotExists) =>
 
       val tableDesc = buildCatalogTable(table, new StructType, partitionCols, bucketSpec,
         properties, provider, options, location, comment, ifNotExists)
       val mode = if (ifNotExists) SaveMode.Ignore else SaveMode.ErrorIfExists
 
       CreateTable(tableDesc, mode, Some(query))
+
+    case create: CreateTableAsSelectStatement =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: shall we implement CREATE TABLE as well?

Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: I thought this PR was getting a little large, so I was going to do create table in a separate one. I can add it to this one if you'd prefer.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala (hunk 1 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: Sure. I'll remove this `orElse` and disable the related tests.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 3]
```diff
@@ -296,6 +298,21 @@ trait CheckAnalysis extends PredicateHelper {
               }
             }
 
+          case CreateTableAsSelect(_, _, partitioning, query, _, _, _) =>
+            val references = partitioning.flatMap(_.references).toSet
+            val badReferences = references.map(_.fieldNames).flatMap { column =>
+              query.schema.findNestedField(column).map(_.dataType) match {
```
[the remark this answers is not in the document]
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: I'll remove it, thanks.

**Comment by rdblue** [comment]
@cloud-fan, I've updated this PR and fixed your review comments, so please have another look. Also, I added a test suite that I accidentally didn't add, `DataSourceV2SQLSuite`. That is where I will be putting end-to-end test cases that use in-memory tables to validate the physical plans.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala (hunk 1 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: This is in #24594.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala [hunk 4]
```diff
@@ -112,4 +129,41 @@ case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with Ca
...
+    val partitioning = ctas.partitioning ++ ctas.bucketSpec.map(_.asTransform)
+
+    // create table properties from TBLPROPERTIES and OPTIONS clauses
+    val properties = new mutable.HashMap[String, String]()
+    properties ++= ctas.properties
+    properties ++= options
+
+    // convert USING, LOCATION, and COMMENT clauses to table properties
+    properties += ("provider" -> ctas.provider)
+    ctas.comment.map(text => properties += ("comment" -> text))
```
Comment (mccheah) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: What happens if these overwrite table properties of the same type in the properties part of the SQL statement itself? Do these overwrite properties of the same name? Are these special property keys documented anywhere?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 5]
```diff
@@ -402,6 +404,35 @@ trait V2WriteCommand extends Command {
...
+    ignoreIfExists: Boolean) extends Command {
+
+  override def children: Seq[LogicalPlan] = Seq(query)
+
+  override lazy val resolved: Boolean = {
+    // the table schema is created from the query schema, so the only resolution needed is to check
+    // that the columns referenced by the table's partitioning exist in the query schema
+    val references = partitioning.flatMap(_.references).toSet
+    references.map(_.fieldNames).forall { column =>
+      query.schema.findNestedField(column) match {
```
Comment (mccheah) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Can we just check on `isDefined` on the returned option?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 6]
```diff
@@ -2031,7 +2031,8 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     if (temporary && ifNotExists) {
       operationNotAllowed("CREATE TEMPORARY TABLE ... IF NOT EXISTS", ctx)
     }
-    (visitTableIdentifier(ctx.tableIdentifier), temporary, ifNotExists, ctx.EXTERNAL != null)
+    val multipartIdentifier: Seq[String] = ctx.multipartIdentifier.parts.asScala.map(_.getText)
```
Comment (mccheah) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Is there a real need to declare the type?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala [hunk 7]
```diff
@@ -47,6 +51,60 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
...
+
+    Utils.tryWithSafeFinallyAndFailureCallbacks({
+      catalog.createTable(ident, query.schema, partitioning.toArray, properties.asJava) match {
+        case table: SupportsWrite =>
+          val builder = table.newWriteBuilder(writeOptions)
+              .withInputDataSchema(query.schema)
+              .withQueryId(UUID.randomUUID().toString)
+          val batchWrite = builder match {
+            case supportsSaveMode: SupportsSaveMode =>
+              supportsSaveMode.mode(SaveMode.Append).buildForBatch()
```
Comment (mccheah) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: I'm not sure why we have to specifically tell the writer to use append mode. Can you elaborate? I think I'm missing something. It would be simpler to remove this branch entirely if possible.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: ah good catch, it should be `query.schema.findNestedField(column).isDefined`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/CreateTablePartitioningValidationSuite.scala [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+      LogicalExpressions.bucket(4, "does_not_exist.z") :: Nil,
+      TestRelation2,
+      Map.empty,
+      Map.empty,
+      ignoreIfExists = false)
+
+    assert(!plan.resolved)
+    assertAnalysisError(plan, Seq(
+      "Invalid partitioning",
+      "does_not_exist.z is missing or is in a map or array"))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/CreateTablePartitioningValidationSuite.scala]: Not a blocker, but we can improve the error message in the future. It's better to let users know which column/field is missing. For example, `a.b`, it's possible that column `a` exists but `a` is not a struct or it doesn't have a `b` field.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala [hunk 9]
```diff
@@ -46,14 +55,22 @@ case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with Ca
...
         properties, provider, options, location, comment, ifNotExists)
       val mode = if (ifNotExists) SaveMode.Ignore else SaveMode.ErrorIfExists
 
       CreateTable(tableDesc, mode, Some(query))
+
+    case create: CreateTableAsSelectStatement =>
+      // the provider was not a v1 source, convert to a v2 plan
+      val CatalogObjectIdentifier(maybeCatalog, identifier) = create.tableName
+      val catalog = maybeCatalog
+          .getOrElse(throw new AnalysisException("Default catalog is not set"))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: nit: `catalog is not specified in the table identifier`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala [hunk 10]
```diff
@@ -112,4 +129,41 @@ case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with Ca
...
       comment = comment)
   }
+
+  private def convertCTAS(
+      catalog: TableCatalog,
+      identifier: Identifier,
+      ctas: CreateTableAsSelectStatement): CreateTableAsSelect = {
+    if (ctas.options.contains("path") && ctas.location.isDefined) {
+      throw new AnalysisException(
+        "LOCATION and 'path' in OPTIONS are both used to indicate the custom table path, " +
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: just for curiosity, does v1 fail this case too?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: This is a good point. When saving Spark tables to Hive metastore, we need to store some spark specific information in the table properties. And we always use `spark.sql.` as the prefix for the property keys.

Shall we follow it and add a prefix here as well?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala (hunk 10 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: and shall we do the same check for `provider` and `comment`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala (hunk 7 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: I think `SupportsSaveMode` is a hack in `TableProvider` only. Seems we don't need to deal with it for `TableCatalog`. Anyway it will be removed soon.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+
+class DataSourceV2SQLSuite extends QueryTest with SharedSQLContext with BeforeAndAfter {
+
+  import org.apache.spark.sql.catalog.v2.CatalogV2Implicits._
+
+  private val orc2 = classOf[OrcDataSourceV2].getName
+
+  before {
+    spark.conf.set("spark.sql.catalog.testcat", classOf[TestInMemoryTableCatalog].getName)
+    spark.conf.set("spark.sql.default.catalog", "testcat")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: I think it has been moved to another PR?

Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: It has. Does this need to be removed? I made the tests that rely on it ignored, so we should just need to revert that. I'd rather keep it so we don't have to figure out why those tests are failing when the other PR is merged.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala (hunk 10 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: Yes, it does fail. And I can add checks for provider and comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala (hunk 9 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: No, this is the correct error message. When no explicit catalog is used, Spark should use the default catalog. When there is no catalog, the problem is that there was no default set.

This is why the default catalog was originally part of this PR. The errors depend on it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 6 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Yes, the compiler was acting strangely and I needed to add it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala (hunk 4 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: I've added validations so that the properties and the clauses cannot both be used. Setting "comment" and using a COMMENT clause will result in an `AnalysisException`.

Passing these as well-known properties was included in the SPIP, but exactly which properties are used should be documented. I'll open a documentation issues and add it as a blocker for the 3.0 release.

If we want to define a prefix for the clauses that are passed as properties, that sounds fine to me. What is a reasonable prefix?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala (hunk 9 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: I'm updating the error message to include both: `No catalog specified for table ... and no default catalog is set`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 6 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: I just took another look and it is no longer needed. Must have been a problem while I hadn't updated everything. I've fixed it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala (hunk 4 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: With the latest changes, conflicting properties are no longer allowed.

Also, I opened https://issues.apache.org/jira/browse/SPARK-27708 to track documentation.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by rdblue** [comment]
Thanks for reviewing, @cloud-fan and @mccheah! I'll get the next few PRs posted.
